keg/generator.py

```diff
@@ -0,0 +1,82 @@
+# Copyright (c) 2020 SUSE Software Solutions Germany GmbH. All rights reserved.
+#
+# This file is part of keg.
+#
+# keg is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# keg is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with keg. If not, see <http://www.gnu.org/licenses/>
+#
+from jinja2 import Environment, FileSystemLoader
+import os
+
+from keg import (
+    image_definition,
+    kiwi_description,
+    template_functions
+)
+from keg.exceptions import KegError, KegKiwiValidationError
+
+
+def create_image_description(image_source,
+                             recipes_root,
+                             dest_dir,
+                             log,
+                             force=False):
+    """
+    Create a KIWI image description in dest_dir from image_source.
+
+    Expects a keg-recipes structure at recipes_root and image_source
+    referring to a directory under images in recipes_root.
```

Comment:
I do not follow why the separation between `image_source` and `recipes_root` is necessary. If `image_source` is expected to be part of `recipes_root`. This need better clarification in the doc and/or code or should be merged into 1 argument. Also `recipes_root` cannot be singular, it has to be a list as we need to be able to build a description from multiple tree definitions by merging. To keep it simple I think we should take a `last one wins` approach and the user is responsible for specifying the order in which things are processed.

Reply:
`recipes_root` specifies where the data resides and `image_source` specifies which image description keg should generate, which refers to a sub directory within the data tree. Those are two distinct input parameters.

Re multiple `recipes_root` directories, I thought we might be able to use git subtree to pull the private bits from a separate repository into the data tree, although I have to admit that I haven't explored git subtrees much yet. I guess which approach to take would also depend to some degree on whether we want to make keg repository-aware or not.

Reply:
I would rather not tie ourselves to a specific RCS, while git is certainly the dominant solution for open source development there are plenty of other options out there and being used.